crates/services/gas_price_service/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+use async_trait::async_trait;
+use fuel_core_services::{
+    RunnableService,
+    RunnableTask,
+    ServiceRunner,
+    StateWatcher,
+};
+use fuel_core_types::fuel_types::BlockHeight;
+use std::sync::Arc;
+
+use tokio::sync::{
+    RwLock,
+    RwLockReadGuard,
+};
+
+pub mod static_updater;
+
+pub fn new_service<A, U>(
+    current_fuel_block_height: BlockHeight,
+    update_algo: U,
+) -> anyhow::Result<ServiceRunner<GasPriceService<A, U>>>
+where
+    U: UpdateAlgorithm<Algorithm = A> + Send + Sync,
+    A: Send + Sync,
+{
+    let service = GasPriceService::new(current_fuel_block_height, update_algo);
+    Ok(ServiceRunner::new(service))
+}
+
+pub struct GasPriceService<A, U> {
+    next_block_algorithm: BlockGasPriceAlgo<A>,
+    update_algorithm: U,
+}
+
+impl<A, U> GasPriceService<A, U>
+where
+    U: UpdateAlgorithm<Algorithm = A>,
+    A: Send + Sync,
+{
+    pub fn new(starting_block_height: BlockHeight, update_algorithm: U) -> Self {
+        let algorithm = update_algorithm.start(starting_block_height);
+        let next_block_algorithm = BlockGasPriceAlgo::new(algorithm);
+        Self {
+            next_block_algorithm,
+            update_algorithm,
+        }
+    }
+
+    pub fn next_block_algorithm(&self) -> BlockGasPriceAlgo<A> {
+        self.next_block_algorithm.clone()
+    }
+}
+
+#[async_trait]
+pub trait UpdateAlgorithm {
```

Comment:
Missing docs. What does this trait exactly do?

Reply:
If I read it correctly, it looks like we should probably use tokio channels for the updating instead of a custom trait.

Reply:
This allows us to customize the behavior of `next`. In our case, we are going to be watching both l2 block commits as well as da recordings. If either of those things happen, we want to update to the `next` gas price algorithm.

Reply:
Ok that makes sense.